wrappers/mongo/mongo.go

```diff
@@ -0,0 +1,504 @@
+		args[1].(interface{}),
+	)
+
+	var document map[string]string
+	decodedResponse := reflect.ValueOf(response).
+		MethodByName("Decode").
+		Call([]reflect.Value{ reflect.ValueOf(&document) })
+	if err := decodedResponse[0]; err.Interface() != nil {
+		logOperationFailure("Could not complete Decode SingleResult", err.String())
+		epsagon.ExtractTracer([]context.Context{}).AddExceptionTypeAndMessage(
+			"mongo-driver",
+			err.String(),
+		)
+	}
+
+	marshalToMetadata(event.Resource.Metadata, "params", args[1])
+	marshalToMetadata(event.Resource.Metadata, "document", document)
+	completeMongoEvent(event)
+	return response
+}
+
+
+func (coll *MongoCollectionWrapper) FindOneAndReplace(args ...interface{}) interface{} {
+	event := startMongoEvent(currentFuncName(), coll)
+	response := coll.collection.FindOneAndReplace(
+		args[0].(context.Context),
+		args[1].(interface{}),
+		args[2].(interface{}),
+	)
+	var document map[string]string
+	decodedResponse := reflect.ValueOf(response).
+		MethodByName("Decode").
+		Call([]reflect.Value{ reflect.ValueOf(&document) })
+	if err := decodedResponse[0]; err.Interface() != nil {
+		logOperationFailure("Could not complete Decode SingleResult", err.String())
+		epsagon.ExtractTracer([]context.Context{}).AddExceptionTypeAndMessage(
+			"mongo-driver",
+			err.String(),
+		)
+	}
+
+	marshalToMetadata(event.Resource.Metadata, "filter", args[1])
+	marshalToMetadata(event.Resource.Metadata, "replacement", args[2])
+	marshalToMetadata(event.Resource.Metadata, "document", document)
+	completeMongoEvent(event)
+	return response
+
+}
+
+
+func (coll *MongoCollectionWrapper) FindOneAndUpdate(args ...interface{}) interface{} {
+	event := startMongoEvent(currentFuncName(), coll)
+	response := coll.collection.FindOneAndReplace(
+		args[0].(context.Context),
+		args[1].(interface{}),
+		args[2].(interface{}),
+	)
+	var document map[string]string
+	decodedResponse := reflect.ValueOf(response).
+		MethodByName("Decode").
+		Call([]reflect.Value{ reflect.ValueOf(&document) })
+	if err := decodedResponse[0]; err.Interface() != nil {
+		logOperationFailure("Could not complete Decode SingleResult", err.String())
+		epsagon.ExtractTracer([]context.Context{}).AddExceptionTypeAndMessage(
+			"mongo-driver",
+			err.String(),
+		)
+	}
+
+	marshalToMetadata(event.Resource.Metadata, "filter", args[1])
+	marshalToMetadata(event.Resource.Metadata, "update", args[2])
+	marshalToMetadata(event.Resource.Metadata, "document", document)
+	completeMongoEvent(event)
+	return response
+}
+
+
+func (coll *MongoCollectionWrapper) Drop(args ...interface{}) error {
+	event := startMongoEvent(currentFuncName(), coll)
+	err := coll.collection.Drop(
+		args[0].(context.Context),
+	)
+	if err != nil {
+		logOperationFailure(fmt.Sprintf("Could not complete %s", currentFuncName()), err.Error())
+		epsagon.ExtractTracer([]context.Context{}).AddExceptionTypeAndMessage(
+			"mongo-driver",
+			err.Error(),
+		)
+	}
+	completeMongoEvent(event)
+	return err
+}
```

Comment:
you should add (for all functions) a deferred func which will call the original handler in case of error where the original handler hasn't been called (due to an error). Use also the  epsagon.GeneralEpsagonRecover as deferred func